text/0000-deprecate-transition-methods-of-controller-and-route.md

```diff
@@ -0,0 +1,253 @@
+for all users of Ember Engines until Ember Engines is refactored to not use the
+deprecated methods anymore.
+
+While this RFC does not intend to decision how Ember Engines should address
+the deprecations, it will very like force Ember Engines to deprecate
+`Route#transitionToExternal`, `Route#replaceWithExternal` and
+`Controller#transitionToExternalRoute` de facto in mid-term for two reasons:
+
+1. These methods would not align anymore with the method provided by Ember to
+   transition between routes.
+2. Ember Engines would not be able to change the implementations of these
+   methods to not use the methods deprecated by this RFC anymore. Using the
+   `RouterService` would very likely change timing and other details of the
+   methods, which could be seen as breaking changes. Keeping the current
+   timings and other details would require usage of private method of the
+   router.
+
+## Alternatives
+
+Three possible alternative are discovered so far:
+
+1. Instead of deprecating `Route#transitionTo`, `Route#replaceWith`,
+   `Controller#transitionToRoute` and `Controller#replaceWith` we could try to
+   align their implementations to match `RouterService#transitionTo` and
+   `RouterService.replaceWith` in regards to timing and other details.
+  
+   Doing so will very likely require breaking changes as existing applications
+   do very likely depend onto the existing timings and other details. Even if
+   not being specified and documented these details became part of our public
+   API over the years.
+  
+   Therefore we would need to introduce an optional feature which allows
+   applications to opt-in into the new timings as soon as they have verifed
+   everything is working as expected.
+
+2. We could introduce new methods on `Route` and `Controller`, which are true
+   shortcuts for using `RouterService`. This would prevent introducing the need
+   to explicitly inject `RouterService` in `Route` and `Controller` in order to
+   trigger a transition. As mentioned above this could be seen as boilerplate.
+
+   Coming up with good names for these new methods would be challening as most
+   of the names has been taken already. Additionally introducing new methods on
+   `Controller` and `Route` will not help with reducing public API surface and
+   teaching a component-service architecture.
+
+3. We could do nothing and continue the different ways to trigger a transition
+   in JavaScript - including different timings and other details.
+
+## Unresolved questions
+
+No open questions have been discovered so far.
+
+## Footnotes
+
+1. Some example for routing related bugs which are open for quite some time
+   now:
+
+   - https://github.com/emberjs/ember.js/issues/10262
+   - https://github.com/emberjs/ember.js/issues/11152
+   - https://github.com/emberjs/ember.js/issues/12945
+   - https://github.com/emberjs/ember.js/issues/14875
+   - https://github.com/emberjs/ember.js/issues/15801
```

Comment:
This issue describes a bug with the Router service's transitionTo method. As a workaround for the issue, you can use the Route/Controller's transitionTo. Is this bug going to be fixed as part of this deprecation? Otherwise, we are stuck between using a deprecated API or having a buggy experience.

Reply:
Yes, we need to fix this before the deprecation goes in. @rwjblue is actively working on this per @krisselden.

Reply:
I came here to mention something similar. There have been multiple cases that we have run into where it's not possible to replace a use of `route.transitionTo` with the `RouterService`'s `transitionTo`. I'll gave to go digging to find the details but they have been related to query params.

I remember at the time I last ran into it reading the [RouterService RFC](https://github.com/emberjs/rfcs/blob/master/text/0095-router-service.md#query-parameter-semantics) and being convinced by the "Query Parameter Semantics" section that this was by design:

> Today, queryParams impose unnecessarily high cost because we cannot generate URLs or determine if a link is active without taking into account the default values of query parameters. Determining their default values is expensive, because it involves instantiating the corresponding controller, even in cases where we will never visit its route.
> Therefore, the queryParams argument to the new `urlFor`, `transitionTo`, `replaceWith`, and `isActive` methods defined in this document will behave differently.

I take that to mean that `RouterService.transitionTo` will never be a drop in replacement for `router.transitionTo`. If this is true then should these edge cases be acknowledged in the "Transition path" section of this RFC?